compiler/src/main/java/com/squareup/anvil/compiler/codegen/dagger/BindsMethodValidator.kt

```diff
@@ -54,70 +79,71 @@ internal class BindsMethodValidator : PrivateCodeGenerator() {
   private fun validateBindsFunction(function: MemberFunctionReference.Psi) {
     if (!function.isAbstract()) {
       throw AnvilCompilationExceptionFunctionReference(
-        message = "@Binds methods must be abstract",
+        message = Errors.BINDS_MUST_BE_ABSTRACT,
         functionReference = function,
       )
     }
 
+    val parameters = function.parameters
+    val psiFunction = function.function
     val hasSingleBindingParameter =
-      (function.parameters.size == 1 && !function.function.isExtensionDeclaration()) ||
-        (function.parameters.isEmpty() && function.function.isExtensionDeclaration())
+      (parameters.size == 1 && !psiFunction.isExtensionDeclaration()) ||
+        (parameters.isEmpty() && psiFunction.isExtensionDeclaration())
```

Comment:
This logic and the new logic below start off by doing the same thing -- trying to figure out what the bound parameter is.  You can simplify your logic to something like:

```kotlin
val bindingParameter = listOfNotNull(
    function.singleParameterTypeOrNull(),
    function.receiverTypeOrNull()
  )
   .singleOrNull()
   ?: throw AnvilCompilationExceptionFunctionReference(...)

val returnTypeName = function.returnTypeOrNull()?.asClassReference()?.asTypeName()
  ?: throw AnvilCompilationExceptionFunctionReference(...)

val paramSuperTypes = bindingParameter.superTypes()

if (returnTypeName !in paramSuperTypes) {
  throw AnvilCompilationExceptionFunctionReference(...)
}
```

Reply:
Yeah, that's much cleaner approach, thanks!